[Mangling] [NFC] Prepare for a new mangling prefix for Embedded Swift: $e

kubamracek opened the pull request:
Given that Embedded Swift's ABI and layout of class metadata is (intentionally) different and incompatible with regular Swift, this PR is introducing a different mangling for all Embedded Swift. Concretely, it changes the mangling prefix from $s to $e.

For reviewing, I recommend looking at commits individually: The first commit is a mechanical change to pass an ASTContext to all instantiations of ASTMangler and its derivatives, which is a large diff but essentially an NFC.

**This PR is a large diff, but it's an NFC, just preparing for an actual mangling change in a subsequent commit.**

kubamracek committed:
[ASTMangler] Pass ASTContext to all instantiations of ASTMangler

kubamracek committed:
[Mangling] Establish a new mangling prefix for Embedded Swift: $e

slavapestov reviewed (commented):
In ASTContext.cpp I see a handful of places where we call getASTContext() on a Type or Decl we have sitting nearby. Do you mind also changing those calls to directly use the new `Context` instance member that you added?

benrimmington commented on include/swift/Demangling/ManglingFlavor.h:
```diff
@@ -0,0 +1,36 @@
+//===--- ManglingUtils.h - Utilities for Swift name mangling ----*- C++ -*-===//
```
```suggestion
//===--- ManglingFlavor.h - Swift name mangling -----------------*- C++ -*-===//
```

benrimmington commented on include/swift/Demangling/ManglingFlavor.h:
```diff
@@ -0,0 +1,36 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
```
```suggestion
// Copyright (c) 2024 Apple Inc. and the Swift project authors
```

benrimmington commented on include/swift/Demangling/ManglingFlavor.h:
```diff
@@ -0,0 +1,36 @@
+
+/// Which mangling style and prefix to use.
+enum class ManglingFlavor: uint8_t {
+  /// Default manling with the ABI stable $s prefix
```
```suggestion
  /// Default mangling with the ABI stable $s prefix
```

benrimmington reviewed (commented):
Should the new prefix be documented? [docs/ABI/Mangling.rst](https://github.com/swiftlang/swift/blob/main/docs/ABI/Mangling.rst)

eeckstein reviewed (commented):
lgtm.
Did you see that a search for `"$s` in the code base finds many hard coded mangled names? For example, in the stdlib we use `@_silgen_name("$s...` a lot. Not sure why.

Another place to update is in swift-demangle: https://github.com/swiftlang/swift/blob/b073d3d186c3dfe173ec8eefa4de87a8860aa506/tools/swift-demangle/swift-demangle.cpp#L341

adrian-prantl reviewed (commented):
I'm missing a change to ` swift/docs/ABI/Mangling.rst `?

adrian-prantl commented:
@augusto2112 If this patch doesn't break any LLDB tests, why?

ahoppen commented:
Could you check if changing the mangling prefix changes the USRs used for indexing? Could you add a test to `test/Index` that builds with embedded mode and just has a FileCheck line for a simple declaration? (`index_curry_thunk.swift` seems to be a good starting point that you can simply further. 

If it does change the USR prefix, I think that’s fine but we might need to do some downstream adjustments for that.

kubamracek committed:
[embedded] Explain new  prefix in docs, fix comments

kubamracek committed:
[Mangling] Also use the new ASTMangler::Context member in ASTMangler.cpp instead of getASTContext() calls

kubamracek commented:
> In ASTContext.cpp I see a handful of places where we call getASTContext() on a Type or Decl we have sitting nearby. Do you mind also changing those calls to directly use the new `Context` instance member that you added?

Addressed.

kubamracek committed:
[embedded] Add an indexer test for Embedded Swift

kubamracek commented:
> I'm missing a change to `swift/docs/ABI/Mangling.rst`?

Added.

kubamracek commented:
> Could you check if changing the mangling prefix changes the USRs used for indexing? Could you add a test to `test/Index` that builds with embedded mode and just has a FileCheck line for a simple declaration? (`index_curry_thunk.swift` seems to be a good starting point that you can simply further.
> 
> If it does change the USR prefix, I think that’s fine but we might need to do some downstream adjustments for that.

Added the test: test/Index/index_embedded.swift.

kubamracek commented:
@swift-ci please test

kubamracek committed:
[embedded] Teach swift-demangle about  prefix, add test

kubamracek commented:
> Another place to update is in swift-demangle:
> 
> https://github.com/swiftlang/swift/blob/b073d3d186c3dfe173ec8eefa4de87a8860aa506/tools/swift-demangle/swift-demangle.cpp#L341

Fixed, added test.

augusto2112 commented on include/swift/RemoteInspection/TypeRefBuilder.h:
```diff
@@ -400,6 +401,10 @@ class TypeRefBuilder {
   TypeRefBuilder &operator=(const TypeRefBuilder &other) = delete;
 
+  Mangle::ManglingFlavor getManglingFlavor() {
+    return Mangle::ManglingFlavor::Default;
```
Can you make this a parameter in the constructor? It can be defaulted to true. Reason being that LLDB uses TypeRefBuilder to process types found in DWARF.

augusto2112 commented on lib/Demangling/Demangler.cpp:
```diff
@@ -185,6 +185,7 @@ int swift::Demangle::getManglingPrefixLength(llvm::StringRef mangledName) {
     /*Swift 4*/   "_T0",
     /*Swift 4.x*/ "$S", "_$S",
     /*Swift 5+*/  "$s", "_$s",
+    /*Swift 5+ Embedded Swift*/  "$e", "_$e",
```
Can you add a function here that returns the Mangling Flavor of a string, and declare it on `Demangle.h` as well?

Something like:
```
std::optional<Mangle::ManglingFlavor> getManglingFlavor(llvm::StringRef mangledName) {
  if (!isSwiftSymbol(mangledName))
    return {};

  if (mangledName.starts_with("$e") || mangledName.starts_with("_$e"))
    return Mangle::ManglingFlavor::Embedded;
  return Mangle::ManglingFlavor::Default;
}
```

kubamracek committed:
[Mangling] Temporarily stage out (#ifdef out) the ASTMangler API changes and Embedded Swift  prefix

kubamracek commented:
@swift-ci please test

kubamracek committed:
[Mangling] Undo test changes for staging of  prefix

kubamracek commented:
@swift-ci please test

kubamracek committed:
[embedded] Annotate test/Index/index_embedded.swift as REQUIRES: swift_feature_Embedded

kubamracek commented:
@swift-ci please test

kubamracek commented:
@swift-ci please test Linux platform

kubamracek commented:
https://github.com/swiftlang/swift-driver/pull/1744

@swift-ci please test
